apm-agent-plugins/apm-kafka-plugin/apm-kafka-headers-plugin/src/main/java/co/elastic/apm/agent/kafka/KafkaProducerHeadersInstrumentation.java

```diff
@@ -94,56 +80,45 @@ public static Span beforeSend(@Advice.FieldValue("apiVersions") final ApiVersion
+                                          @Advice.Argument(value = 0) ProducerRecord record,
+                                          @Advice.Argument(value = 1) Callback callback,
+                                          @Advice.This final KafkaProducer thiz,
+                                          @Advice.Thrown @Nullable final Throwable throwable) {
+
+            Span span = enter != null ? (Span) enter[0] : null;
+            if (span == null) {
+                return null;
+            }
+            Object[] overrideThrowable = null;
             if (throwable != null && throwable.getMessage().contains("Magic v1 does not support record headers")) {
                 // Probably our fault - ignore span and retry. May happen when using a new client with an old (< 0.11.0)
                 // broker. In such cases we DO check the version, but the first version check may be not yet up to date.
-                if (span != null) {
-                    logger.info("Adding header to Kafka record is not allowed with the used broker, attempting to resend record");
-                    //noinspection unchecked
-                    record = new ProducerRecord(record.topic(), record.partition(), record.timestamp(),
-                        record.key(), record.value(), record.headers());
-                    //noinspection ConstantConditions
-                    KafkaInstrumentationHeadersHelper<ConsumerRecord, ProducerRecord> kafkaInstrumentationHelper =
-                        kafkaInstrHeadersHelperManager.getForClassLoaderOfClass(KafkaProducer.class);
-                    if (kafkaInstrumentationHelper != null) {
-                        kafkaInstrumentationHelper.removeTraceContextHeader(record);
-                    }
-                    span.deactivate();
-                    span = null;
-                    headersSupported = false;
-                    return true;
-                }
-            }
-            //noinspection ConstantConditions
-            KafkaInstrumentationHelper<Callback, ProducerRecord, KafkaProducer> helper = kafkaInstrHelperManager.getForClassLoaderOfClass(KafkaProducer.class);
-            if (helper != null && span != null) {
-                helper.onSendEnd(span, record, thiz, throwable);
+                logger.info("Adding header to Kafka record is not allowed with the used broker, attempting to resend record");
+                ProducerRecord copy = new ProducerRecord<>(record.topic(), record.partition(), record.timestamp(),
+                    record.key(), record.value(), record.headers());
+                headersHelper.removeTraceContextHeader(copy);
+                headersSupported = false;
+                thiz.send(copy, callback);
+                overrideThrowable = new Object[]{null};
             }
-            return false;
+            helper.onSendEnd(span, record, thiz, throwable);
+            return overrideThrowable;
```

Comment:
[question] Here, could we just return `Throwable` to override the exception that we might have triggered ? Do we need to return an `Object[]` here, even if just with a single value ?

If I understand correctly, the behavior from the returned value here is
- `null` : don't change the thrown exception if there is any
- `Object[]{null}` : remove the exception
- `Object[]{ new Exception() }` : override/set the exception to throw

To me, returning an array would only be required in the unlikely case where we try to set the return value or a class attribute, in which case we would have to always return the full array with the original exception we got from `@Advice.Thrown` argument.

Reply:
That doesn't work because we can only remove the exception with `Object[]{null}`.
When changing the return type to `Throwable` and returning `null`, it would still not change the exception if there's any.
That is to be consistent with the `@AssignTo.*` annotations that don't allow assigning something to `null`. The reason is that if an advice method throws an exception (which we suppress), the return value will always be null. We want to avoid assigning that null value as that will have unintended consequences.